airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/datetime_stream_state_converter.py

```diff
@@ -0,0 +1,190 @@
+            "slices": slices,
+            "legacy": stream_state,
+        }
+
+    def convert_to_sequential_state(self, cursor_field: CursorField, stream_state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        """
+        Convert the state message from the concurrency-compatible format to the stream's original format.
+
+        e.g.
+        { "created": "2021-01-18T21:18:20.000Z" }
+        """
+        if self.is_state_message_compatible(stream_state):
+            legacy_state = stream_state.get("legacy", {})
+            if slices := stream_state.pop("slices", None):
+                latest_complete_time = self._get_latest_complete_time(slices)
+                if latest_complete_time:
+                    legacy_state.update({cursor_field.cursor_field_key: self.output_format(latest_complete_time)})
+            return legacy_state or {}
+        else:
+            return stream_state
+
+    def _get_latest_complete_time(self, slices: List[MutableMapping[str, Any]]) -> Optional[datetime]:
+        """
+        Get the latest time before which all records have been processed.
+        """
+        if slices:
+            first_interval = self.merge_intervals(slices)[0][self.END_KEY]
+            return first_interval
+        else:
+            return None
+
+
+class EpochValueConcurrentStreamStateConverter(DateTimeStreamStateConverter):
+    """
+    e.g.
+    { "created": 1617030403 }
+    =>
+    {
+        "state_type": "date-range",
+        "metadata": { … },
+        "slices": [
+            {starts: 0, end: 1617030403, finished_processing: true}
+        ]
+    }
+    """
+
+    _zero_value = 0
+
+    def increment(self, timestamp: datetime) -> Any:
+        return timestamp + timedelta(seconds=1)
+
+    def output_format(self, timestamp: datetime) -> int:
+        return int(timestamp.timestamp())
+
+    def parse_timestamp(self, timestamp: int) -> datetime:
+        dt_object = pendulum.from_timestamp(timestamp)
+        if not isinstance(dt_object, DateTime):
+            raise ValueError(f"DateTime object was expected but got {type(dt_object)} from pendulum.parse({timestamp})")
+        return dt_object  # type: ignore  # we are manually type checking because pendulum.parse may return different types
+
+
+class IsoMillisConcurrentStreamStateConverter(DateTimeStreamStateConverter):
+    """
+    e.g.
+    { "created": "2021-01-18T21:18:20.000Z" }
+    =>
+    {
+        "state_type": "date-range",
+        "metadata": { … },
+        "slices": [
+            {starts: "2020-01-18T21:18:20.000Z", end: "2021-01-18T21:18:20.000Z", finished_processing: true}
+        ]
+    }
+    """
+
+    _zero_value = "0001-01-01T00:00:00.000Z"
+
+    def increment(self, timestamp: datetime) -> Any:
+        return timestamp + timedelta(milliseconds=1)
+
+    def output_format(self, timestamp: datetime) -> Any:
+        return timestamp.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
```

Comment:
nit: if the format and the timedelta were object fields, we might be able to re-use the same code. I'm not sure if that's true for the epoch timestamp since `%s` is not a [recommended format directive](https://stackoverflow.com/a/11743262)